subsys/shell/backends/shell_rpmsg.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2024 Basalte bv
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/shell/shell_rpmsg.h>
+
+SHELL_RPMSG_DEFINE(shell_transport_rpmsg);
+SHELL_DEFINE(shell_rpmsg, CONFIG_SHELL_PROMPT_RPMSG, &shell_transport_rpmsg,
+	     CONFIG_SHELL_BACKEND_RPMSG_LOG_MESSAGE_QUEUE_SIZE,
+	     CONFIG_SHELL_BACKEND_RPMSG_LOG_MESSAGE_QUEUE_TIMEOUT, SHELL_FLAG_OLF_CRLF);
+
+static int rpmsg_shell_cb(struct rpmsg_endpoint *ept, void *data,
+			  size_t len, uint32_t src, void *priv)
+{
+	const struct shell_transport *transport = (const struct shell_transport *)priv;
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+	struct shell_rpmsg_rx rx;
+
+	if (len == 0) {
+		return RPMSG_ERR_NO_BUFF;
+	}
+
+	rx.data = data;
+	rx.len = len;
+	if (k_msgq_put(&sh_rpmsg->rx_q, &rx, K_NO_WAIT) != 0) {
+		return RPMSG_ERR_NO_MEM;
+	}
+
+	rpmsg_hold_rx_buffer(ept, data);
+	sh_rpmsg->shell_handler(SHELL_TRANSPORT_EVT_RX_RDY, sh_rpmsg->shell_context);
+
+	return RPMSG_SUCCESS;
+}
+
+static int uninit(const struct shell_transport *transport)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+
+	if (!sh_rpmsg->ready) {
+		return -ENODEV;
+	}
+
+	rpmsg_destroy_ept(&sh_rpmsg->ept);
+	sh_rpmsg->ready = false;
+
+	return 0;
+}
+
+static int init(const struct shell_transport *transport,
+		const void *config,
+		shell_transport_handler_t evt_handler,
+		void *context)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+	struct rpmsg_device *rdev;
+	int ret;
+
+	if (sh_rpmsg->ready) {
+		return -EALREADY;
+	}
+
+	if (config == NULL) {
+		return -EINVAL;
+	}
+	rdev = (struct rpmsg_device *)config;
+
+	k_msgq_init(&sh_rpmsg->rx_q, (char *)sh_rpmsg->rx_buf, sizeof(struct shell_rpmsg_rx),
+		    CONFIG_SHELL_RPMSG_MAX_RX);
+
+	ret = rpmsg_create_ept(&sh_rpmsg->ept, rdev, CONFIG_SHELL_RPMSG_SERVICE_NAME,
+			       CONFIG_SHELL_RPMSG_SRC_ADDR, CONFIG_SHELL_RPMSG_DST_ADDR,
+			       rpmsg_shell_cb, NULL);
+	if (ret < 0) {
+		return ret;
+	}
+
+	sh_rpmsg->ept.priv = (void *)transport;
+
+	sh_rpmsg->shell_handler = evt_handler;
+	sh_rpmsg->shell_context = context;
+	sh_rpmsg->ready = true;
+
+	return 0;
+}
+
+static int enable(const struct shell_transport *transport, bool blocking)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+
+	ARG_UNUSED(blocking);
+
+	if (!sh_rpmsg->ready) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int write(const struct shell_transport *transport,
+		 const void *data, size_t length, size_t *cnt)
+{
+	struct shell_rpmsg *sh_rpmsg = (struct shell_rpmsg *)transport->ctx;
+	int ret;
+	uint32_t write_len;
+	void *tx_buf;
+
+	*cnt = 0;
+
+	if (!sh_rpmsg->ready) {
+		return -ENODEV;
+	}
+
+	tx_buf = rpmsg_get_tx_payload_buffer(&sh_rpmsg->ept, &write_len, false);
```

Comment:
I had a look to the freeze. The issue is that no more buffer can be allocated at this point ( not yet free by Linux). In such case the write returns an error.
By updating with following code, no more issue on my side.
```c
	while (!tx_buf) {
		tx_buf = rpmsg_get_tx_payload_buffer(&sh_rpmsg->ept, &write_len, false);
		if (tx_buf == NULL)
			k_sleep(K_USEC(5000));
	}
```
could you retry `rpmsg_send `? The function sould handle the way of free buffer

Reply:
Hey @arnopo, I've given this some more thought, and it feels like **waiting** is similar to **blocking**, and we have an API function to make the shell backend blocking. I think we need to adhere to the blocking setting by using `rpmsg_send`/`rpmsg_trysend` accordingly.

It would be nice to have some "events" to know if a `tty` device driver has been opened/closed (rpmsg driver layer) or if TX buffers become available (lower layer?).

Crazy idea: add an `rpmsg-tty-txt` compatible to the linux driver and send `STX` and `ETX` ascii characters on open and close. _But I have no idea how the linux drive works._
This would also add the option to flush input buffers on `EOT` characters.

Reply:
> Hey @arnopo, I've given this some more thought, and it feels like **waiting** is similar to **blocking**, and we have an API function to make the shell backend blocking. I think we need to adhere to the blocking setting by using `rpmsg_send`/`rpmsg_trysend` accordingly.

Yes my fix is just for test, this should not be the final fix. 
Notice that the `rpmsg_send `bloack using `metal_sleep_usec` that calls `k_sleep`. So the thread should be suspended.

> 
> It would be nice to have some "events" to know if a `tty` device driver has been opened/closed (rpmsg driver layer) or if TX buffers become available (lower layer?).

Today when we create the endpoint on Zephyr we generate a name service announcement that creates the `/dev/ttyRPMSG` on Linux side.
There is also possibility to do it (with one patch in linux) to initiate the TTY from linux.
For details see demo 3 [here](https://openamp.readthedocs.io/en/latest/openamp-system-reference/examples/zephyr/rpmsg_multi_services/README.html). in such case there is the `rpmsg_ns_unbind_cb` that can be used to detect device creation.

Concerning the TX buffers, depending on the platform implementation, the Linux sends a mailbox notif when it frees a buffer. But could be tricky to manage as rpmsg buffers are shared between several rpmsg services.

> 
> Crazy idea: add an `rpmsg-tty-txt` compatible to the linux driver and send `STX` and `ETX` ascii characters on open and close. _But I have no idea how the linux driver works._ This would also add the option to flush input buffers on `EOT` characters.

I have started some time ago a mechnism to manage the flow control. but unfortunatly, no time to go back on it:
https://lore.kernel.org/lkml/e54bcfcb-8e37-9caa-b330-a7411820b7ce@foss.st.com/T/